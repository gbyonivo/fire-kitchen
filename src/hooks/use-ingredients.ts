import {
  setFetchingIngredients,
  setFetchingIngredientsError,
  setIngredients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngredients(): {
  ingredients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngredients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingredients,
    fetchingIngredients,
    fetchingIngredientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngredients = useCallback(() => {
    dispatch(setFetchingIngredients());
    const fetch = async () => {
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngredients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngredientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngredientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingredients.length > 0 || fetchingIngredients) return;
    fetchIngredients();
  }, [fetchIngredients, ingredients, fetchingIngredients]);

  return {
    ingredients,
    fetchIngredients,
    fetching: fetchingIngredients,
    error: fetchingIngredientsError,
    ingridientCache,
  };
}
