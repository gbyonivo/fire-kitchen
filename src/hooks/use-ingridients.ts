import {
  setFetchingIngridients,
  setFetchingIngridientsError,
  setIngridients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngridients(): {
  ingridients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngridients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingridients,
    fetchingIngridients,
    fetchingIngridientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngridients = useCallback(() => {
    console.log("fetching ingridients 1");
    dispatch(setFetchingIngridients());
    const fetch = async () => {
      console.log("fetching ingridients 2");
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngridients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngridientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngridientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingridients.length > 0 || fetchingIngridients) return;
    console.log("fetching ingridients");
    fetchIngridients();
  }, [fetchIngridients, ingridients, fetchingIngridients]);

  return {
    ingridients,
    fetchIngridients,
    fetching: fetchingIngridients,
    error: fetchingIngridientsError,
    ingridientCache,
  };
}
