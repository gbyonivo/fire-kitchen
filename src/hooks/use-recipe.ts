import { RootState } from "@/lib/store";
import { useCallback, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { AxiosError } from "axios";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { updateRecipeCache } from "@/lib/slices/cache-slice";

export function useRecipe({ recipeId }: { recipeId: string }) {
  const dispatch = useDispatch();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<AxiosError | null>(null);

  const recipeCache = useSelector(
    (state: RootState) => state.cache.recipeCache
  );
  const recipe = recipeCache?.[recipeId];

  const fetchRecipe = useCallback(() => {
    const getRecipe = async () => {
      if (recipe) return;
      try {
        setLoading(true);
        const response = await KitchenAxios.get(`/lookup.php`, {
          params: {
            i: recipeId,
          },
        });
        const recipe = convertRecipe(response.data.meals[0]);
        if (recipe) {
          dispatch(updateRecipeCache(recipe));
        } else {
          setError(new AxiosError("Recipe not found"));
        }
        setLoading(false);
      } catch (error) {
        setLoading(false);
        setError(error as AxiosError);
      }
    };
    getRecipe();
  }, [dispatch, recipeId, recipe]);

  useEffect(() => {
    if (recipe?.id || loading || error) return;
    fetchRecipe();
  }, [fetchRecipe, recipe?.id, loading, error]);

  return {
    recipe,
    fetchRecipe,
    loading,
    error,
  };
}
