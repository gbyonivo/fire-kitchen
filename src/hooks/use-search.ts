import { searchSuccess, setSearching } from "@/lib/slices/search-slice";
import { Ingredient } from "@/types/ingredient";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { useCallback } from "react";
import { useDispatch } from "react-redux";

export function useSearch() {
  const dispatch = useDispatch();

  const searchRecipe = useCallback(
    async (ingredients: Ingredient[], search: string) => {
      try {
        dispatch(setSearching());
        const searches = [];
        if (ingredients.length > 0) {
          searches.push(
            KitchenAxios.get("filter.php", {
              params: {
                i: ingredients
                  .map((ingredient) => ingredient.ingredient)
                  .join(","),
              },
            })
          );
        }
        if (search.length > 2) {
          searches.push(
            KitchenAxios.get("/search.php", {
              params: {
                s: search,
              },
            })
          );
        }

        const responses = await Promise.all(searches);
        const recipes = responses.flatMap(
          (response) => response.data.meals || []
        );
        const uniqueRecipes = recipes.reduce((acc, recipe) => {
          if (!recipe?.idMeal) {
            return acc;
          }
          return {
            ...acc,
            [recipe.idMeal]: convertRecipe(recipe),
          };
        }, {});

        console.log(recipes.length, Object.values(uniqueRecipes));

        dispatch(searchSuccess(Object.values(uniqueRecipes)));
      } catch (error) {
        console.error(error);
      }
    },
    [dispatch]
  );

  return { searchRecipe };
}
