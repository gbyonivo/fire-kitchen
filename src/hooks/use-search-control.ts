import { useRef, useState } from "react";

import { useRouter, useSearchParams } from "next/navigation";

import { usePathname } from "next/navigation";
import { useSearch } from "./use-search";
import { createQueryString } from "@/utils/common";
import { useEffect } from "react";
import { Ingredient } from "@/types/ingredient";
import { useIngredients } from "./use-ingredients";

export const useSearchControl = () => {
  const pathname = usePathname();
  const router = useRouter();
  const searchParams = useSearchParams();
  const { ingridientCache } = useIngredients();
  const timeout = useRef(null);
  const { searchRecipe } = useSearch();
  const getIngredients = () => {
    return searchParams
      .get("ingredients")
      ?.split(",")
      .reduce((acc: Ingredient[], ingredient) => {
        if (ingridientCache?.[ingredient]) {
          return [...acc, ingridientCache?.[ingredient]];
        }
        return acc;
      }, []);
  };

  console.log(getIngredients());

  const [usedParams, setUsedParams] = useState<{
    search: string;
    ingredients: Ingredient[];
  }>(() => ({
    search: searchParams.get("search") || "",
    ingredients: getIngredients(),
  }));

  useEffect(() => {
    router.push(
      pathname +
        "?" +
        createQueryString({
          nameValuePairs: [
            { name: "search", value: usedParams.search },
            {
              name: "ingredients",
              value: usedParams.ingredients.map((i) => i.ingredient).join(","),
            },
          ],
          searchParams,
        })
    );
    searchRecipe(usedParams.ingredients, usedParams.search);
  }, [pathname, router, searchParams, searchRecipe, usedParams]);

  return {
    onSetParams: (params: { search: string; ingredients: Ingredient[] }) => {
      setUsedParams(params);
    },
    initialSearchValue: searchParams.get("search") || "",
    initialIngredients: getIngredients(),
  };
};
