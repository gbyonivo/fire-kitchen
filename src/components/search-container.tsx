"use client";

import { Button } from "./common/button";
import { MultiSelect } from "./common/multi-select";
import { useState } from "react";
import { useIngredients } from "@/hooks/use-ingredients";
import { Ingredient } from "@/types/ingredient";
import { TextInput } from "./common/text-input";
import { useSelector } from "react-redux";
import { RootState } from "@/lib/store";
import { RecipeCard } from "./common/recipe-card";
import { Spinner } from "./common/spinner";
import Link from "next/link";
import { useSearchControl } from "@/hooks/use-search-control";

export function SearchContainer() {
  const { ingredients } = useIngredients();
  const { onSetParams, initialIngredients, initialSearchValue } =
    useSearchControl();
  const { recipes, searching } = useSelector(
    (state: RootState) => state.search
  );
  const [selectedIngredients, setSelectedIngredients] = useState<Ingredient[]>(
    () => initialIngredients
  );
  const [search, setSearch] = useState(() => initialSearchValue);

  return (
    <div className="flex flex-col">
      <div className="pt-8 flex flex-col lg:flex-row gap-2 justify-center">
        <MultiSelect<Ingredient>
          options={ingredients}
          selectedOptions={selectedIngredients}
          onChange={(ingredients) => setSelectedIngredients(ingredients)}
          getLabel={(ingredient) => ingredient.ingredient}
          getSearchKey={(ingredient) => ingredient.ingredient}
          getKey={(ingredient) => ingredient.ingredient}
          containerClassName="lg:w-96"
          label="Ingredients"
        />
        <TextInput
          value={search}
          onChange={(value) => setSearch(value)}
          containerClassName="lg:w-64"
          label="Keyword (at least 2 characters)"
          disabled={searching}
        />
        <Button
          className="self-start mt-2 lg:mt-6"
          onClick={() =>
            onSetParams({ search, ingredients: selectedIngredients })
          }
          disabled={
            searching || (search.length < 2 && !selectedIngredients.length)
          }
          id="search-button"
        >
          Search
        </Button>
      </div>
      <div className="flex justify-center items-center h-16">
        {searching && <Spinner size="large" />}
      </div>
      <div className="flex-1 grid grid-cols-2 gap-2 md:gap-16 md:grid-cols-4 lg:grid-cols-6 py-8">
        {recipes.map((recipe) => {
          return (
            <Link
              href={`/recipes/${recipe.id}`}
              key={recipe.id}
              id={`recipe-link-${recipe.id}`}
            >
              <RecipeCard recipe={recipe} id={`recipe-card-${recipe.id}`} />
            </Link>
          );
        })}
      </div>
    </div>
  );
}
