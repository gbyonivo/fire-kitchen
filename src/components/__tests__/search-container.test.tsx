// import { render } from "@testing-library/react";
import { useSelector } from "react-redux";
import { useIngredients } from "@/hooks/use-ingredients";
import { useSearchControl } from "@/hooks/use-search-control";
import { render, screen, waitFor } from "@testing-library/react";
import { SearchContainer } from "../search-container";
import userEvent from "@testing-library/user-event";
// import { SearchContainer } from "../search-container";

jest.mock("@/hooks/use-ingredients");
jest.mock("@/hooks/use-search-control");
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
}));

// add more tests when there is result on the screen

describe("SearchContainer", () => {
  test.each([
    [{ initialIngredients: [], initialSearchValue: "Chicken", called: true }],
    [{ initialIngredients: [], initialSearchValue: "", called: false }],
  ])(
    "should match snapshot and onSetParams should be called when search button is clicked",
    async ({ initialIngredients, initialSearchValue, called }) => {
      const onSetParams = jest.fn();
      (useSelector as unknown as jest.Mock).mockReturnValue({
        recipes: [],
        searching: false,
      });
      (useIngredients as unknown as jest.Mock).mockReturnValue({
        ingredients: [],
        ingridientCache: {},
      });
      (useSearchControl as unknown as jest.Mock).mockReturnValue({
        onSetParams,
        initialIngredients,
        initialSearchValue,
      });
      const { container } = render(<SearchContainer />);
      await userEvent.click(screen.getByTestId("search-button"));
      await waitFor(() => {
        if (called) {
          expect(onSetParams).toHaveBeenCalled();
        } else {
          expect(onSetParams).not.toHaveBeenCalled();
        }
      });
      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    }
  );

  test("should match snapshot and button should be disabled when searching", async () => {
    const onSetParams = jest.fn();
    (useSelector as unknown as jest.Mock).mockReturnValue({
      recipes: [],
      searching: true,
    });
    (useIngredients as unknown as jest.Mock).mockReturnValue({
      ingredients: [],
      ingridientCache: {},
    });
    (useSearchControl as unknown as jest.Mock).mockReturnValue({
      onSetParams,
      initialIngredients: [],
      initialSearchValue: "",
    });
    expect(1).toBe(1);
    const { container } = render(<SearchContainer />);

    await userEvent.click(screen.getByTestId("search-button"));
    await waitFor(() => {
      expect(onSetParams).not.toHaveBeenCalled();
    });
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
