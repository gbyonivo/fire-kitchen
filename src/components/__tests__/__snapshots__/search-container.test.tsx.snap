// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer should match snapshot and button should be disabled when searching 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="pt-8 flex flex-col lg:flex-row gap-2 justify-center"
    >
      <div
        class="lg:w-96"
      >
        <label
          class="block text-sm/6 font-medium text-gray-400"
          data-headlessui-state=""
          for="headlessui-listbox-button-«rn»"
          id="headlessui-label-«rm»"
        >
          Ingredients
        </label>
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="headlessui-label-«rm» headlessui-listbox-button-«rn»"
          class="relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10 focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
          data-headlessui-state=""
          id="headlessui-listbox-button-«rn»"
          type="button"
        >
          <div
            class="flex flex-wrap gap-2"
          />
          <svg
            aria-hidden="true"
            class="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="lg:w-64"
      >
        <div
          data-headlessui-state=""
        >
          <label
            class="block text-sm/6 font-medium text-gray-400"
            data-headlessui-state=""
            for="headlessui-control-«rr»"
            id="headlessui-label-«rs»"
          >
            Keyword (at least 2 characters)
          </label>
          <input
            aria-labelledby="headlessui-label-«rs»"
            class="h-10 block w-full rounded-lg border-none bg-white/5 px-3 py-1.5 text-sm/6 text-white focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
            data-disabled=""
            data-headlessui-state="disabled"
            disabled=""
            id="headlessui-control-«rr»"
            value=""
          />
          <span
            hidden=""
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
          />
        </div>
      </div>
      <button
        class="inline-flex items-center gap-2 rounded-md 
        px-3 py-1.5 text-sm/6 font-semibold  shadow-inner 
        shadow-white/10 focus:not-data-focus:outline-none data-focus:outline 
        data-focus:outline-white  bg-gray-600 data-open:bg-gray-600 text-gray-400 self-start mt-2 lg:mt-6"
        data-disabled=""
        data-headlessui-state="disabled"
        data-testid="search-button"
        disabled=""
        type="button"
      >
        <div
          class="flex items-center gap-2"
        >
          Search
        </div>
      </button>
    </div>
    <div
      class="flex justify-center items-center h-16"
    >
      <div>
        <svg
          aria-hidden="true"
          class="animate-spin text-red-500 undefined w-16 h-16"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.963 2.286a.75.75 0 0 0-1.071-.136 9.742 9.742 0 0 0-3.539 6.176 7.547 7.547 0 0 1-1.705-1.715.75.75 0 0 0-1.152-.082A9 9 0 1 0 15.68 4.534a7.46 7.46 0 0 1-2.717-2.248ZM15.75 14.25a3.75 3.75 0 1 1-7.313-1.172c.628.465 1.35.81 2.133 1a5.99 5.99 0 0 1 1.925-3.546 3.75 3.75 0 0 1 3.255 3.718Z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div
      class="flex-1 grid grid-cols-2 gap-2 md:gap-16 md:grid-cols-4 lg:grid-cols-6 py-8"
    />
  </div>
</div>
`;

exports[`SearchContainer should match snapshot and onSetParams should be called when search button is clicked 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="pt-8 flex flex-col lg:flex-row gap-2 justify-center"
    >
      <div
        class="lg:w-96"
      >
        <label
          class="block text-sm/6 font-medium text-gray-400"
          data-headlessui-state=""
          for="headlessui-listbox-button-«r3»"
          id="headlessui-label-«r2»"
        >
          Ingredients
        </label>
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="headlessui-label-«r2» headlessui-listbox-button-«r3»"
          class="relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10 focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
          data-headlessui-state=""
          id="headlessui-listbox-button-«r3»"
          type="button"
        >
          <div
            class="flex flex-wrap gap-2"
          />
          <svg
            aria-hidden="true"
            class="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="lg:w-64"
      >
        <div
          data-headlessui-state=""
        >
          <label
            class="block text-sm/6 font-medium text-gray-400"
            data-headlessui-state=""
            for="headlessui-control-«r7»"
            id="headlessui-label-«r8»"
          >
            Keyword (at least 2 characters)
          </label>
          <input
            aria-labelledby="headlessui-label-«r8»"
            class="h-10 block w-full rounded-lg border-none bg-white/5 px-3 py-1.5 text-sm/6 text-white focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
            data-headlessui-state=""
            id="headlessui-control-«r7»"
            value="Chicken"
          />
          <span
            hidden=""
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
          />
        </div>
      </div>
      <button
        class="inline-flex items-center gap-2 rounded-md 
        px-3 py-1.5 text-sm/6 font-semibold  shadow-inner 
        shadow-white/10 focus:not-data-focus:outline-none data-focus:outline 
        data-focus:outline-white  bg-gray-800 data-hover:bg-gray-700 data-open:bg-gray-800 text-white self-start mt-2 lg:mt-6"
        data-focus=""
        data-headlessui-state="focus"
        data-testid="search-button"
        type="button"
      >
        <div
          class="flex items-center gap-2"
        >
          Search
        </div>
      </button>
    </div>
    <div
      class="flex justify-center items-center h-16"
    />
    <div
      class="flex-1 grid grid-cols-2 gap-2 md:gap-16 md:grid-cols-4 lg:grid-cols-6 py-8"
    />
  </div>
</div>
`;

exports[`SearchContainer should match snapshot and onSetParams should be called when search button is clicked 2`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="pt-8 flex flex-col lg:flex-row gap-2 justify-center"
    >
      <div
        class="lg:w-96"
      >
        <label
          class="block text-sm/6 font-medium text-gray-400"
          data-headlessui-state=""
          for="headlessui-listbox-button-«rd»"
          id="headlessui-label-«rc»"
        >
          Ingredients
        </label>
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="headlessui-label-«rc» headlessui-listbox-button-«rd»"
          class="relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10 focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
          data-headlessui-state=""
          id="headlessui-listbox-button-«rd»"
          type="button"
        >
          <div
            class="flex flex-wrap gap-2"
          />
          <svg
            aria-hidden="true"
            class="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="lg:w-64"
      >
        <div
          data-headlessui-state=""
        >
          <label
            class="block text-sm/6 font-medium text-gray-400"
            data-headlessui-state=""
            for="headlessui-control-«rh»"
            id="headlessui-label-«ri»"
          >
            Keyword (at least 2 characters)
          </label>
          <input
            aria-labelledby="headlessui-label-«ri»"
            class="h-10 block w-full rounded-lg border-none bg-white/5 px-3 py-1.5 text-sm/6 text-white focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
            data-headlessui-state=""
            id="headlessui-control-«rh»"
            value=""
          />
          <span
            hidden=""
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
          />
        </div>
      </div>
      <button
        class="inline-flex items-center gap-2 rounded-md 
        px-3 py-1.5 text-sm/6 font-semibold  shadow-inner 
        shadow-white/10 focus:not-data-focus:outline-none data-focus:outline 
        data-focus:outline-white  bg-gray-600 data-open:bg-gray-600 text-gray-400 self-start mt-2 lg:mt-6"
        data-disabled=""
        data-headlessui-state="disabled"
        data-testid="search-button"
        disabled=""
        type="button"
      >
        <div
          class="flex items-center gap-2"
        >
          Search
        </div>
      </button>
    </div>
    <div
      class="flex justify-center items-center h-16"
    />
    <div
      class="flex-1 grid grid-cols-2 gap-2 md:gap-16 md:grid-cols-4 lg:grid-cols-6 py-8"
    />
  </div>
</div>
`;
